staging/src/k8s.io/apiserver/pkg/util/shufflesharding/shufflesharding_test.go

```diff
@@ -272,33 +275,73 @@ func TestShuffleAndDealToSliceWithValidation(t *testing.T) {
 
 func BenchmarkShuffleAndDealToSlice(b *testing.B) {
 	hashValueBase := math.MaxUint64 / uint64(b.N)
-	deckSize, handSize := int32(512), int32(8)
+	deckSize, handSize := 512, 8
 	for i := 0; i < b.N; i++ {
 		_ = ShuffleAndDealToSlice(hashValueBase*uint64(i), deckSize, handSize)
 	}
 }
 
+// ff computes the falling factorial `n!/(n-m)!` and requires n to be
+// positive and m to be in the range [0, n] and requires the answer to
+// fit in an int
+func ff(n, m int) int {
+	ans := 1
+	for f := n; f > n-m; f-- {
+		ans *= f
+	}
+	return ans
+}
+
 func TestUniformDistribution(t *testing.T) {
-	deckSize, handSize := int32(128), int32(3)
-	handCoordinateMap := make(map[int]int)
+	const spare = 64 - maxHashBits
+	tests := []struct {
+		deckSize, handSize int
+		hashMax            int
+	}{
+		{64, 3, 1 << uint(math.Ceil(math.Log2(float64(ff(64, 3))))+spare)},
+		{128, 3, ff(128, 3)},
+		{128, 3, 3 * ff(128, 3)},
+		{70, 4, ff(70, 4)},
+	}
+	for _, test := range tests {
+		handCoordinateMap := make(map[int]int) // maps coded hand to count of times seen
 
-	allCoordinateCount := 128 * 127 * 126 / 6
+		fallingFactorial := ff(test.deckSize, test.handSize)
+		permutations := ff(test.handSize, test.handSize)
+		allCoordinateCount := fallingFactorial / permutations
+		nff := float64(test.hashMax) / float64(fallingFactorial)
+		minCount := permutations * int(math.Floor(nff))
+		maxCount := permutations * int(math.Ceil(nff))
+		aHand := make([]int, test.handSize)
+		for i := 0; i < test.hashMax; i++ {
```

Comment:
I don't think it is good to test from 0 to hashMax, because we all know the principle of our algorithm, it will get same results in [allCoordinateCount*N, allCoordinateCount*(N+1)), so the test case will always succeed, which is not expected.

Reply:
I think the point of unit tests is to check whether the implementation is doing what it is intended to do.  If we were confident that the implementation does what it is intended to do then these unit tests would prove nothing, and the only things to prove would be math exercises that do not need to be repeated in unit tests.

By exploring from 0 to hashMax-1, we can test a range of hash space that is like what gets used in the priority and fairness filter --- a range from 0 to SomePowerOfTwo-1.

Reply:
OK, I think these test cases are enough for priority and fairness filter.